### PR TITLE
feat: support other audio format

### PR DESCRIPTION
**Pull request by AnneYang720**
Goals:

This PR is related to issue #998 
- [x] refactor `load()` method in AudioUrl to use PyDub
- [x] remove useless code
- [x] pass current tests
- [x] add tests for other format

### CODE REVIEW

File: docarray/typing/url/audio_url.py [hunk 1]
```diff
@@ -66,9 +66,16 @@ class MyDoc(Document):
...
 
-        bytes_ = AudioBytes(self.load_bytes())
-        return bytes_.load()
+        bytes_ = self.load_bytes()
+        segment = AudioSegment.from_file(BytesIO(bytes_))
+
+        samples = segment.get_array_of_samples()
+        samples = np.array(samples)
+
+        samples_norm = samples / 2 ** (segment.sample_width * 8 - 1)
```
Comment (JohannesMessner) [inline docarray/typing/url/audio_url.py:77]: I have no idea about audio data so maybe this is a stupid question, but what are these numbers here? Are they guaranteed to be always these same numbers, no matter the data? And what does this transformation do?

**Review by JohannesMessner — CHANGES_REQUESTED** [review]

File: docarray/typing/url/audio_url.py (hunk 1 above)
Comment (AnneYang720) [reply docarray/typing/url/audio_url.py:77]: Line 74 and 75 are how to get ndarray from audio file with pydub AudioSegment.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -210,6 +214,7 @@ jobs:
           poetry install --all-extras
           poetry run pip install protobuf==3.19.0
           poetry run pip install tensorflow==2.11.0
+          sudo apt-get install --no-install-recommends ffmpeg
```
Comment (samsja) [inline .github/workflows/ci.yml:217]: `--no-install-recommends` what is this doing ?

Comment (AnneYang720) [reply .github/workflows/ci.yml:217]: Don't let apt to install some "recommended" (not needed) packages, to save a little bit of CI time.

**Review by samsja — APPROVED** [review]

File: docarray/typing/url/audio_url.py [hunk 3]
```diff
@@ -66,9 +44,14 @@ class MyDoc(Document):
...
 
         """
+        from pydub import AudioSegment  # type: ignore
+
+        bytes_ = self.load_bytes()
+        segment = AudioSegment.from_file(BytesIO(bytes_))
+        samples = np.array(segment.get_array_of_samples())
 
-        bytes_ = AudioBytes(self.load_bytes())
-        return bytes_.load()
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/url/audio_url.py:71]: good point

[the remark this answers is not in the document]
Comment (AnneYang720) [reply docarray/typing/url/audio_url.py:71]: The problem is `AudioBytes` and its `load()` method are not currently used anywhere.

**Comment by AnneYang720** [comment]
I added the function `save_to_audio_file`. During the development, I found the `frame_rate` of an audio file is essential, so I added an variable `frame_rate` for `class Audio`. So now it is
```python
audio = Audio(
    url='https://github.com/docarray/docarray/blob/feat-rewrite-v2/tests/toydata/hello.wav?raw=true'
)
audio.tensor, audio.frame_rate = audio.url.load()
```
just like in video we have
```python
vid.audio.tensor, vid.tensor, vid.key_frame_indices = vid.url.load()
```

It may not be very compact. Maybe I can wrap the return to a class like `VideoLoadResult`.

File: docarray/typing/tensor/audio/abstract_audio_tensor.py [hunk 4]
```diff
@@ -43,6 +43,28 @@ def save_to_wav_file(
             f.setframerate(sample_rate)
             f.writeframes(self.to_bytes())
 
+    def save_to_audio_file(
```
[the remark this answers is not in the document]
Comment (samsja) [reply docarray/typing/tensor/audio/abstract_audio_tensor.py:46]: yes

**Review by samsja — CHANGES_REQUESTED** [review]
add sample rate to audio predefined doc

File: docarray/typing/url/audio_url.py (hunk 1 above)
Comment (JohannesMessner) [reply docarray/typing/url/audio_url.py]: okk I see. Maybe we should have a comment like `# normalize amplitude to [-1, 1]`?

**Comment by samsja** [comment]
> 
    * simply have a huge-ass list of all supported file extensions, move it to a different file if needed
    * if a path has an extension and it is not in that list, then fail
    * if a path does not have an extension, pass
  
I like this tradeoff. 

The goal of validation is mostly to catch if someone put an `file.png` instead of a `file.wav`.

File: docarray/typing/tensor/audio/abstract_audio_tensor.py [hunk 5]
```diff
@@ -20,28 +19,36 @@ def to_bytes(self):
...
 
-    def save_to_wav_file(
+    def save(
         self: 'T',
         file_path: Union[str, BinaryIO],
-        sample_rate: int = 44100,
+        format: str = 'wav',
+        frame_rate: int = 44100,
         sample_width: int = 2,
+        *args,
```
Comment (samsja) [inline docarray/typing/tensor/audio/abstract_audio_tensor.py:28]: I know we had this pattern a lot in v1, but I am not sure i lile it tbh.

What about having a `pydub_args: Dict[str, Any]` parameters

and do segment.export(..., **pydub_arg)

**Review by samsja — CHANGES_REQUESTED** [review]
looks great, one comment about the save function

File: docarray/typing/tensor/audio/abstract_audio_tensor.py (hunk 5 above)
Comment (samsja) [reply docarray/typing/tensor/audio/abstract_audio_tensor.py:28]: @JohannesMessner what do u think ?

**Review by samsja — APPROVED** [review]

File: docarray/typing/tensor/audio/abstract_audio_tensor.py [hunk 6]
```diff
@@ -20,28 +19,35 @@ def to_bytes(self):
...
 
-    def save_to_wav_file(
+    def save(
         self: 'T',
         file_path: Union[str, BinaryIO],
-        sample_rate: int = 44100,
+        format: str = 'wav',
+        frame_rate: int = 44100,
         sample_width: int = 2,
+        pydub_args: Dict[str, Any] = {},
```
Comment (samsja) [inline docarray/typing/tensor/audio/abstract_audio_tensor.py:28]: can u add the corresponding docstring as well ? Just say that this arg are forwarded to the export function

**Review by samsja — APPROVED** [review]

**Review by JohannesMessner — APPROVED** [review]

**Pull request merged**